### PR TITLE
Add ORTModelForVision2Seq for VisionEncoderDecoder models inference

### PR DESCRIPTION
**Pull request by mht-sharma**
# What does this PR do?

This PR enables the inference of the `VisionEncoderDecoder` models using ONNXRuntime. The PR adds `ORTModelForVision2Seq` for doing inference similar to `AutoModelForVision2Seq` by changing just a few lines.

# Usage

```diff
>>> from PIL import Image
>>> from transformers import GPT2TokenizerFast, ViTImageProcessor
->>>  from transformers import AutoModelForVision2Seq
+>>>  from optimum.onnxruntime import ORTModelForVision2Seq
>>> import requests

>>> model_name = "nlpconnect/vit-gpt2-image-captioning"
->>> model = AutoModelForVision2Seq.from_pretrained(model_name)
+>>> model = ORTModelForVision2Seq.from_pretrained(model_name, from_transformers=True)
>>> tokenizer = GPT2TokenizerFast.from_pretrained(model_name)
>>> image_processor = ViTImageProcessor.from_pretrained(model_name)

>>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
>>> image = Image.open(requests.get(url, stream=True).raw)
>>> pixel_values = image_processor(image, return_tensors="pt").pixel_values

>>> generated_ids = model.generate(pixel_values)
>>> generated_text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0]
>>> print(generated_text)
```

## Limitations
- `Donut` model not supported
- `TrOCR` model not supported with `use_cache=True`.
-  IObinding not supported

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 1]
```diff
@@ -24,18 +24,20 @@
...
 
 import torch
-from transformers import AutoModelForSeq2SeqLM, AutoModelForSpeechSeq2Seq, GenerationConfig
+from transformers import AutoModelForSeq2SeqLM, AutoModelForSpeechSeq2Seq, AutoModelForVision2Seq, GenerationConfig
 from transformers.file_utils import add_start_docstrings_to_model_forward
 from transformers.modeling_outputs import BaseModelOutput, Seq2SeqLMOutput
 
 import onnxruntime as ort
 from huggingface_hub import hf_hub_download
+from onnxruntime import InferenceSession
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:33]: Can you actually remove `import onnxruntime as ort` and `from onnxruntime import InferenceSession` and just put `import onnxruntime as ort` inside the `TYPE_CHECKING` block? And quote everything in the rest of the code. I think it is used only for type hints.

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 2]
```diff
@@ -79,6 +81,12 @@
             Mel features extracted from the raw speech waveform. `(batch_size, feature_size, encoder_sequence_length)`.
 """
 
+VISION_ENCODER_INPUTS_DOCSTRING = r"""
+    Args:
+        pixel_values (`torch.FloatTensor`):
+            Featuers extracted from an Image. `(batch_size, num_channels, height, width)`.
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:87]:
```suggestion
            Features extracted from an image. This tensor should be of shape `(batch_size, num_channels, height, width)`.
```

It's just RGB images no? Not "feature" in the sense of any preprocessing has been applied yet?

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 3]
```diff
@@ -130,6 +138,22 @@
             `(batch_size, num_heads, encoder_sequence_length, embed_size_per_head)`.
 """
 
+VISION_ENCODER_DECODER_SEQ2SEQ_ONNX_MODEL_DOCSTRING = r"""
+    Args:
+        pixel_values (`torch.FloatTensor`):
+            Featuers extracted from an Image.
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:144]:
```suggestion
            Features extracted from an image.
```

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 4]
```diff
@@ -130,6 +138,22 @@
...
+            `(batch_size, num_channels, height, width)`.
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:145]: same

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 5]
```diff
@@ -130,6 +138,22 @@
...
+        decoder_input_ids (`torch.LongTensor`):
+            Indices of decoder input sequence tokens in the vocabulary of shape `(batch_size, decoder_sequence_length)`.
+        encoder_outputs (`torch.FloatTensor`):
+            The encoder `last_hidden_state` of shape `(batch_size, encoder_sequence_length, hidden_size)`.
+        past_key_values (`tuple(tuple(torch.FloatTensor), *optional*)`
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:150]:
```suggestion
        past_key_values (`tuple(tuple(torch.FloatTensor), *optional*)`
```

We should not put `*optional*` in Optimum following this discussion: https://huggingface.slack.com/archives/C02P0559X9S/p1669628754896019

Can rather precise what the default is? Although, for the type: `Tuple[Tuple[torch.FloatTensor]]`

(edit: this is copy paste, so could you edit it as well in the rest?)

File: tests/onnxruntime/test_modeling.py [hunk 6]
```diff
@@ -2742,6 +2747,188 @@ def test_compare_generation_to_io_binding(self, test_name: str, model_arch: str,
...
+    }
+
+    ORTMODEL_CLASS = ORTModelForVision2Seq
+
+    TASK = "vision2seq-lm"
+
+    def exclude_trocr_with_cache(params):
+        if params[0] == "trocr" and params[1] == True:
+            return None
+        return params
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:2766]: Just for my knowledge, why is this not supported? What will happen if an user tries to use `ORTModelForVision2Seq` with `trocr` with `use_cache = True`?

File: optimum/utils/normalized_config.py [hunk 7]
```diff
@@ -166,7 +173,6 @@ class NormalizedConfigManager:
         'roformer',
         'segformer',
         'squeezebert',
-        'vit',
```
Comment (fxmarty) [inline optimum/utils/normalized_config.py:169]: you can remove `deit` as well above

**Review by fxmarty — APPROVED** [review]
LGTM as long as the tests pass, thank you for the addition! Could you also make sure that the code snippet in the documentation is valid?

File: optimum/onnxruntime/base.py [hunk 8]
```diff
@@ -115,39 +115,40 @@ def __init__(
...
-        self.expected_value_symbolic_shape = None
-        for output in self.session.get_outputs():
-            if ".key" in output.name:
-                self.expected_key_symbolic_shape = output.shape
-            elif ".value" in output.name:
-                self.expected_value_symbolic_shape = output.shape
-            # To handle the old case when past_key_values were following the format: past_key_values_{idx}
-            elif "key_values" in output.name:
-                if self.expected_key_symbolic_shape is None:
+        if self.parent_model.use_cache is True or len(self.key_value_output_names) != 0:
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/base.py:118]:
```suggestion
        if len(self.key_value_output_names) != 0:
```
Since the next if raises an error if not?

**Review by michaelbenayoun — APPROVED** [review]
LGTM!

What's the status for the pat key / values?

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 2 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: I think the normalization and resize would be applied to the image

Comment (fxmarty) [reply optimum/onnxruntime/modeling_seq2seq.py]: right!

File: tests/onnxruntime/test_modeling.py (hunk 6 above)
Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py:2766]: The model currently not output any `past_key_values` when `use_cache=True`.  So need to figure out why. Probably something wrong in the modeling code.

```
| What will happen if an user tries to use ORTModelForVision2Seq with trocr with use_cache = True?
```

For this I have added an error message during export.

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 5 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: Added default to None as in the function sig

**Comment by mht-sharma** [comment]
> LGTM!
> 
> What's the status for the pat key / values?

Past key values is not supported only for TrOCR, would look into this after Donut

File: optimum/onnxruntime/base.py (hunk 8 above)
Comment (mht-sharma) [reply optimum/onnxruntime/base.py:118]: This I added for `TrOCR` because when use_cache = False, currently it does not output the `past_key_values`. So this will skip this and user can do inference with use_cache = False
